pkg/gateway/client/group.go

```
@@ -0,0 +1,318 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/obot-platform/obot/pkg/accesstoken"
+	"github.com/obot-platform/obot/pkg/auth"
+	"github.com/obot-platform/obot/pkg/gateway/types"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+const (
+	// groupCheckPeriod defines how often the system checks for updates to group information from the auth provider.
+	groupCheckPeriod = time.Minute * 10
+)
+
+// ListAuthGroups lists the auth provider groups for the given auth provider.
+//
+// It supports fuzzy finding group names using on the given nameFilter.
+// It queries the auth provider for "live" group search from the auth provider, then combines the
+// results with cached groups from the database.
+// This allows admins to discover groups that authenticated users belong to for auth providers
+// limited group search capabilities; e.g. there's not an effective way to perform a fuzzy search for
+// GitHub teams or orgs by name.
+func (c *Client) ListAuthGroups(ctx context.Context, authProviderURL, authProviderNamespace, authProviderName, nameFilter string) ([]types.Group, error) {
+	// Fetch groups from the auth provider
+	providerGroups := []auth.GroupInfo{}
+	if authProviderURL != "" {
+		u, err := url.Parse(authProviderURL + "/obot-list-auth-groups")
+		if err == nil {
+			if nameFilter != "" {
+				q := u.Query()
+				q.Set("name", nameFilter)
+				u.RawQuery = q.Encode()
+			}
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+			if err == nil {
+				if accessToken := accesstoken.GetAccessToken(ctx); accessToken != "" {
+					req.Header.Set("Authorization", "Bearer "+accessToken)
+				}
+
+				resp, err := http.DefaultClient.Do(req)
+				if err == nil {
+					defer resp.Body.Close()
+					if resp.StatusCode == http.StatusOK {
+						_ = json.NewDecoder(resp.Body).Decode(&providerGroups)
```

Why do we want to ignore this error?